### ISSUE TITLE
MPI All-to-all internal error

### ISSUE THREAD

**Issue by MonkD3** [pmodels/mpich#6708]
# Environment

- **System:** MeluXina supercomputer using the CPU partition. Detailed architecture can be found [here](https://docs.lxp.lu/system/detailed_arch/). CPU partition is also called Cluster module.
- **MPICH version:** 4.1.2 compiled from source with GCC 11.3.0 and the following flags :
```console
./mpiexec --version
HYDRA build details:
    Version:                                 4.1.2
    Release Date:                            Wed Jun  7 15:22:45 CDT 2023
    CC:                              gcc
    Configure options:                       '--disable-option-checking' '--with-hwloc=embedded' '--enable-g=most' '--disable-tag-error-bits' '--with-device=ch4:ucx' '--with-ucx=<path of UCX> '--enable-threads=single' '--disable-cxx' 'CC=gcc' 'CXX=g++' 'FC=gfortran' '--cache-file=/dev/null' '--srcdir=.' 'CFLAGS= -g -O2' 'LDFLAGS= <link paths>'
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Demux engines available:                 poll select
```
UCX is taken from the available module on the cluster, version 1.13.1.

# Minimal working example

Here is a minimal reproduction of the error :
```c
#include <mpi.h>
#include <stdlib.h>
#include <stdio.h>

int main(int argc, char** argv){
    MPI_Init(&argc, &argv);
    int world_size, rank;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    int* sendbuf = (int*) malloc(sizeof(int) * world_size);
    int* recvbuf = (int*) malloc(sizeof(int) * world_size);
    for (int i = 0; i < world_size; i++){
        sendbuf[i] = rank;
        recvbuf[i] = 0;
    }

    int retval = MPI_Alltoall(sendbuf, 1, MPI_INT, recvbuf, 1, MPI_INT, MPI_COMM_WORLD);
    if (retval != MPI_SUCCESS) {
        printf("ERROR: alltoall failed\n");
        MPI_Finalize();
        return EXIT_FAILURE;
    }

    free(sendbuf);
    free(recvbuf);
    MPI_Finalize();
    return EXIT_SUCCESS;
}
```

This is then compiled and run on the compute node with the batch script:
```bash
#!/bin/bash -l
#SBATCH --nodes=48
#SBATCH --cpus-per-task=1
#SBATCH --tasks-per-node=128

mpicc bug.c
mpiexec -n 6144 -ppn 128 ./a.out
```
Which produce the following bug :
```console
Abort(344006158) on node 6016 (rank 6016 in comm 0): Fatal error in internal_Alltoall: Message truncated, error stack:
internal_Alltoall(129).......................: MPI_Alltoall(sendbuf=0x1a0e7a0, sendcount=1, MPI_INT, recvbuf=0x1a147b0, recvcount=1, MPI_INT, MPI_COMM_WORLD) failed
MPID_Alltoall(1070)..........................:
MPIDI_Alltoall_allcomm_composition_json(1012):
MPIDI_Alltoall_intra_composition_beta(1035)..:
MPIR_Alltoall_impl(3046).....................:
MPIR_Alltoall_allcomm_auto(2970).............:
MPIR_Alltoall_intra_brucks(118)..............:
(unknown)(): Message truncated
```

## Additional informations 

Note that using 64 nodes and 8192 ranks produces the same issue. On the other hand, using less then 32 full nodes (128 ppn) works without issue.

Also note that the non-blocking version of this call (`MPI_Ialltoall`) does not produce issues.

Exporting the environment variables `MPIR_CVAR_DEVICE_COLLECTIVES=none` or `MPIR_CVAR_ALLTOALL_INTRA_ALGORITHM=brucks` does not change the outcome.

# TL;DR

There seem to be an internal error in the all-to-all blocking call when calling with 48 or 64 full nodes (128 ppn) in MPICH 4.1.2.

**Comment by raffenet** [comment]
Sorry for the lack of update. I've tracked this down to an issue with the way `ch4:ucx` handles noncontig datatypes at the receiver. An internal message in the brucks algo sends packed data to a noncontig receive buffer. UCX is fragmenting the message without knowledge of the basic datatype boundaries. When UCX tries to unpack the data via the generic datatype callback we supplied, some bytes are leftover and causes an error.

For now, I think we will patch ch4:ucx to always receive noncontig data into a temporary buffer and unpack at completion. Otherwise we risk this issue for any contig->noncontig message.

**Comment by MonkD3** [comment]
Great, thanks for the update. 
Am I correct to assume that this will be included in the next release ?

**Comment by raffenet** [comment]
Yes. I need to make some changes to the PR. Once accepted, it will be part of 4.2.0 and likely backported to a 4.1.3 release.
